File: rust/checks/doc_required.py
"""Enforce /// doc comments on public Rust items — rust/docs.md

REQUIRED on:
  pub fn, pub struct, pub enum, pub trait, pub type, pub mod, pub const/static
  (including pub(crate) and pub(super))

EXEMPT:
  - test files (path contains /tests/ or /test/)
  - pub use re-exports
"""

from __future__ import annotations

import re
from pathlib import Path
from typing import Generator

from rulestools.issue import Issue, Severity

_RULE = "rust/docs/doc-required"

_PUB_ITEM = re.compile(
    r"^\s*pub(?:\([^)]+\))?\s+(?:async\s+)?(?:fn|struct|enum|trait|type|mod|const|static)\s+(\w+)"
)
_ATTRIBUTE = re.compile(r"^\s*#\[")
_PUB_USE   = re.compile(r"^\s*pub\s+use\b")


def _is_test_file(path: Path) -> bool:
    parts = path.parts
    return "tests" in parts or "test" in parts


def _has_doc_comment(lines: list[str], item_idx: int) -> bool:
    """Walk backwards skipping attributes; return True if /// is found."""
    i = item_idx
    while i > 0:
        i -= 1
        trimmed = lines[i].strip()
        if not trimmed:
            return False
        if trimmed.startswith("///"):
            return True
        if _ATTRIBUTE.match(lines[i]):
            continue  # skip attribute lines
        return False
    return False


def check(path: Path, lines: list[str]) -> Generator[Issue, None, None]:
    if _is_test_file(path):
        return

    for idx, line in enumerate(lines):
        if _PUB_USE.match(line):
            continue
        m = _PUB_ITEM.match(line)
        if not m:
            continue
        name = m.group(1)
        if not _has_doc_comment(lines, idx):
            yield Issue(
                file=path,
                line=idx + 1,
                col=1,
                severity=Severity.ERROR,
                rule=_RULE,
                message=f"pub item `{name}` is missing a `///` doc comment — add one above the declaration",
            )

File: crates/scanner/src/checks/doc_required.rs
use std::path::Path;

use regex::Regex;
use std::sync::LazyLock;

use crate::config::Config;
use crate::context::FileContext;
use crate::issue::{Issue, Severity};

static PUB_ITEM_RE: LazyLock<Regex> = LazyLock::new(|| {
    Regex::new(r"^\s*pub(?:\([^)]+\))?\s+(?:async\s+)?(?:fn|struct|enum|trait|type|mod|const|static)\s+(\w+)").unwrap()
});

/// Check that all pub items have `///` doc comments.
pub fn check(
    file_ctx: &FileContext,
    lines: &[&str],
    _cfg: &Config,
    issues: &mut Vec<Issue>,
    path: &Path,
) {
    if file_ctx.is_test_file {
        return;
    }

    for (line_num, line) in lines.iter().enumerate() {
        // Skip `pub use` re-exports
        if line.contains(" use ") {
            continue;
        }

        if let Some(caps) = PUB_ITEM_RE.captures(line) {
            let item_name = &caps[1];
            if !has_doc_comment(lines, line_num) {
                issues.push(Issue::new(
                    path,
                    line_num + 1,
                    1,
                    Severity::Error,
                    "rust/docs/doc-required",
                    format!("pub item `{item_name}` is missing a `///` doc comment"),
                ));
            }
        }
    }
}

/// Walk backwards from `line_idx`, skipping `#[...]` attributes, looking for `///`.
fn has_doc_comment(lines: &[&str], line_idx: usize) -> bool {
    if line_idx == 0 {
        return false;
    }
    let mut idx = line_idx - 1;
    loop {
        let trimmed = lines[idx].trim();

        if trimmed.starts_with("///") {
            return true;
        }
        // Skip attributes like #[derive(...)], #[cfg(...)], etc.
        if trimmed.starts_with("#[") {
            if idx == 0 {
                return false;
            }
            idx -= 1;
            continue;
        }
        // Any other line means no doc comment
        return false;
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::context::Language;

    fn prod_ctx() -> FileContext {
        FileContext { language: Language::Rust, is_test_file: false, is_mother_file: false, is_definition_file: false }
    }

    #[test]
    fn documented_fn_ok() {
        let mut issues = Vec::new();
        let lines = vec!["/// Does something.", "pub fn foo() {}"];
        check(&prod_ctx(), &lines, &Config::default(), &mut issues, Path::new("a.rs"));
        assert!(issues.is_empty());
    }

    #[test]
    fn undocumented_fn_flagged() {
        let mut issues = Vec::new();
        let lines = vec!["", "pub fn foo() {}"];
        check(&prod_ctx(), &lines, &Config::default(), &mut issues, Path::new("a.rs"));
        assert_eq!(issues.len(), 1);
        assert!(issues[0].message.contains("foo"));
    }

    #[test]
    fn doc_through_attributes() {
        let mut issues = Vec::new();
        let lines = vec!["/// Documented.", "#[derive(Debug)]", "pub struct Foo;"];
        check(&prod_ctx(), &lines, &Config::default(), &mut issues, Path::new("a.rs"));
        assert!(issues.is_empty());
    }

    #[test]
    fn pub_use_skipped() {
        let mut issues = Vec::new();
        let lines = vec!["pub use crate::other::Thing;"];
        check(&prod_ctx(), &lines, &Config::default(), &mut issues, Path::new("a.rs"));
        assert!(issues.is_empty());
    }
}
